### ISSUE TITLE
[Bug] Non-determinism found during fuzzing

### ISSUE THREAD

**Issue by cretz** [temporalio/sdk-dotnet#330]
### Describe the bug

Given the [KitchenSinkWorkflow.cs.txt](https://github.com/user-attachments/files/16599943/KitchenSinkWorkflow.cs.txt) workflow and [KitchenSink.cs.txt](https://github.com/user-attachments/files/16599955/KitchenSink.cs.txt) proto model (may need to adjust namespace), running the [omes-replay.json](https://github.com/user-attachments/files/16599958/omes-replay.json) history on it via a replayer gives:

> Temporalio.Exceptions.WorkflowNondeterminismException : Nondeterminism: Workflow activation completion failed: Failure { failure: Some(Failure { message: "[TMPRL1100] Nondeterminism error: Activity machine does not handle this event: HistoryEvent(id: 1182, StartChildWorkflowExecutionInitiated)", source: "", stack_trace: "", encoded_attributes: None, cause: None, failure_info: Some(ApplicationFailureInfo(ApplicationFailureInfo { r#type: "", non_retryable: false, details: None, next_retry_delay: None })) }), force_cause: NonDeterministicError }

This history was obtained via https://github.com/temporalio/omes and this repo, both checked out at the time of this writing, and running this command with a dev server running:

    go run ./cmd run-scenario-with-worker --scenario fuzzer --language cs --option seed=17951219391681344745 --iterations 1 --version /path/to/sdk-dotnet

Each run may give a different history.

We are still investigating this. Debugging how the original run may have given a mismatching history is challenging.

**Comment by cretz** [comment]
Here is another one history from the same scenario but with cache/sticky disabled: [omes-replay-2.json](https://github.com/user-attachments/files/16600152/omes-replay-2.json) that causes:

> Temporalio.Exceptions.WorkflowNondeterminismException : Nondeterminism: Workflow activation completion failed: Failure { failure: Some(Failure { message: "[TMPRL1100] Nondeterminism error: Timer machine does not handle this event: HistoryEvent(id: 624, ActivityTaskScheduled)", source: "", stack_trace: "", encoded_attributes: None, cause: None, failure_info: Some(ApplicationFailureInfo(ApplicationFailureInfo { r#type: "", non_retryable: false, details: None, next_retry_delay: None })) }), force_cause: NonDeterministicError }

**Comment by cretz** [comment]
Here's a simple test to add to SDK to replicate via replayer:

```csharp
public class OmesTest : WorkflowEnvironmentTestBase
{
    public OmesTest(ITestOutputHelper output, WorkflowEnvironment env)
        : base(output, env)
    {
    }

    [Fact]
    public async Task ReplayWorkflowAsync_NonDeterministicRunFromOmes_Fails()
    {
        // Replay history from JSON
        var history = WorkflowHistory.FromJson(
            "w-nespuagc-1",
            TestUtils.ReadAllFileText("omes-replay.json"));

        var replayer = new WorkflowReplayer(
            new WorkflowReplayerOptions()
            {
                LoggerFactory = LoggerFactory,
            }.AddWorkflow<KitchenSinkWorkflow>());
        await replayer.ReplayWorkflowAsync(history);
    }
}
```

**Comment by dandavison** [comment]
Here's another example from the fuzzer just in case it's useful. Sorry, the oss-cicd pipeline that caught this is not yet in a position to save and retrieve the fuzzer proto.

> {
  "message": "[TMPRL1100] Nondeterminism error: Child workflow machine does not handle this event:HistoryEvent { event_id: 1709, event_time: Some(Timestamp { seconds: 1724331693, nanos: 159859858 }), event_type: WorkflowExecutionUpdateCompleted, version: 0, task_id: 1061367, worker_may_ignore: false, user_metadata: None, attributes: Some(WorkflowExecutionUpdateCompletedEventAttributes(WorkflowExecutionUpdateCompletedEventAttributes { meta: Some(Meta { update_id: \"a90bbbb8-10ca-4356-955a-66c58155cc42\", identity: \"7@oss-cicd-worker-dd844b6d4-kvh5f@\" }), accepted_event_id: 1692, outcome: Some(Outcome { value: Some(Success(Payloads { payloads: [Payload { metadata: {\"messageType\": "temporal.omes.kitchen_sink.WorkflowState", \"encoding\": "json/protobuf"}, data: "{ "kvs": { "1": "x", "10": "x", "12": "x", "14": "x", "2": "x", "21": "x", "22": "x", "24": "x", "29": "x", "3": "x", "34": "x", "36": "x", "40": "x", "41": "x", "45": "x", "48": "x", "53": "x", "57": "x", "59": "x", "61": "x", "72": "x", "83": "x", "84": "x", "85": "x", "89": "x", "91": "x", "93": "x" } }" }] })) }) })) }",
  "applicationFailureInfo": {}
}

**Comment by cretz** [comment]
Yeah, it's quite easy to replicate now, but I am struggling mightily to debug.

**Comment by cretz** [comment]
After digging, this has been replicated and is believed to be a core issue. https://github.com/temporalio/sdk-core/issues/803 has been opened.

**Comment by cretz** [comment]
This should now be fixed with the 1.3.1 release.
